File: ncServicos/rCrediario.cs
using System;
using System.Transactions;
using ncNComum;
using ncRepositorios;
using ncModelos;

namespace ncServicos.nsCrediario
{
    public class rCrediario
    {
        private readonly IpCrediario _crediario;
        private readonly IpParcela _parcela;

        public rCrediario(IpCrediario crediario, IpParcela parcela)
        {
            _crediario = crediario;
            _parcela = parcela;
        }

        public ColecaoCrediario Listar()
        {
            try { return _crediario.Listar(); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Listar Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public int ConsultarMax()
        {
            try { return _crediario.ConsultarMax(); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarMax Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoCrediario Consultar(dCrediario dados)
        {
            try { return _crediario.Consultar(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Consultar Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoParcelas ConsultarParcelas(dParcelas dados)
        {
            try { return _parcela.Consultar(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarParcelas Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoParcelas ConsultarParcelasPagamentos(dParcelas dados)
        {
            try { return _parcela.ConsultarPagamentos(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarParcelasPagamentos Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public dParcelas ConsultarParcela(int cid)
        {
            try
            {
                var lista = _parcela.Consultar(new dParcelas { cid = cid });
                return lista != null && lista.Count > 0 ? lista[0] : null;
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarParcela Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoParcelas ConsultarParcelasCliente(int codCliente)
        {
            try { return _parcela.ConsultarParcelasCliente(codCliente); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarParcelasCliente Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoParcelas ConsultarParcelasVencidas(int codCliente)
        {
            try { return _parcela.ConsultarParcelasVencidas(codCliente); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarParcelasVencidas Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public int Incluir(dCrediario dados, ColecaoParcelas dadosParcelas)
        {
            try
            {
                using (var ts = new TransactionScope())
                {
                    int retorno = _crediario.Incluir(dados);
                    foreach (var parcela in dadosParcelas)
                    {
                        parcela.crediarioId = retorno;
                        _parcela.Incluir(parcela);
                    }
                    ts.Complete();
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Incluir Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public int IncluirCrediario(dCrediario dados)
        {
            try { return _crediario.Incluir(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em IncluirCrediario [" + ToString() + "] - " + ex.Message); }
        }

        public int IncluirParcela(dParcelas dadosParcela)
        {
            try { return _parcela.Incluir(dadosParcela); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em IncluirParcela Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public int Alterar(dCrediario dados, ColecaoParcelas dadosParcelas)
        {
            try
            {
                using (var ts = new TransactionScope())
                {
                    int retorno = _crediario.Alterar(dados);
                    foreach (var parcela in dadosParcelas)
                        _parcela.Alterar(parcela);
                    ts.Complete();
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Alterar Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public int GravarPagamentoParcelas(ColecaoParcelas dadosParcelas)
        {
            try
            {
                using (var ts = new TransactionScope())
                {
                    foreach (var parcela in dadosParcelas)
                        _parcela.IncluirPagamento(parcela);
                    ts.Complete();
                    return 1;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em GravarPagamentoParcelas Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public int Renegociar(dCrediario dados, ColecaoParcelas dadosParcelas)
        {
            try
            {
                using (var ts = new TransactionScope())
                {
                    int retorno = dados.cid != 0 ? _crediario.Alterar(dados) : _crediario.Incluir(dados);
                    _crediario.ExcluirParcelas(dados);
                    foreach (var parcela in dadosParcelas)
                    {
                        if (parcela.crediarioId == 0)
                            parcela.crediarioId = retorno;
                        _parcela.Incluir(parcela);
                    }
                    ts.Complete();
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Renegociar Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public int AlterarCrediario(dCrediario dados)
        {
            try { return _crediario.Alterar(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em AlterarCrediario [" + ToString() + "] - " + ex.Message); }
        }

        public int AlterarControle(dCrediario dados)
        {
            try { return _crediario.AlterarControle(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em AlterarControle Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public int Excluir(dCrediario dados)
        {
            try { return _crediario.Excluir(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Excluir Crediario [" + ToString() + "] - " + ex.Message); }
        }

        public int CorrigirParcelas()
        {
            try { return _parcela.Corrigir(); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em CorrigirParcelas Crediario [" + ToString() + "] - " + ex.Message); }
        }
    }
}

File: ncRegras/rCrediario.vb
Imports System.Transactions

Imports ncDados.nsCrediario
Imports ncPersistencia.nsCrediario
Imports ncComum.nsFuncoes
Imports ncComum.nsExcecao
Imports ncComum.nsLog.cLog

Namespace nsCrediario

    Public Class rCrediario

        '-- Métodos de controle ( Várias chamadas; Controle de transação )

        Public Function Listar() As ColecaoCrediario

            Dim retorno As ColecaoCrediario
            Dim persistencia As pCrediario
            Dim retornoPersistencia As ColecaoCrediario

            Try

                retorno = New ColecaoCrediario

                persistencia = New pCrediario
                retornoPersistencia = persistencia.Listar()

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Listar Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Listar = retorno

        End Function

        Public Function ConsultarMax() As Integer

            Dim retorno As Integer
            Dim persistencia As pCrediario

            Try

                persistencia = New pCrediario
                retorno = persistencia.ConsultarMax()

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em ConsultarMax Venda [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            ConsultarMax = retorno

        End Function

        Public Function Consultar(ByVal dados As dCrediario) As ColecaoCrediario

            Dim retorno As ColecaoCrediario
            Dim persistencia As pCrediario
            Dim retornoPersistencia As ColecaoCrediario

            Try

                retorno = New ColecaoCrediario

                persistencia = New pCrediario
                retornoPersistencia = persistencia.Consultar(dados)

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function ConsultarParcelas(ByVal dados As dParcelas) As ColecaoParcelas

            Dim retorno As ColecaoParcelas
            Dim persistencia As pParcela
            Dim retornoPersistencia As ColecaoParcelas

            Try

                retorno = New ColecaoParcelas

                persistencia = New pParcela
                retornoPersistencia = persistencia.Consultar(dados)

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function ConsultarParcelasPagamentos(ByVal dados As dParcelas) As ColecaoParcelas

            Dim retorno As ColecaoParcelas
            Dim persistencia As pParcela
            Dim retornoPersistencia As ColecaoParcelas

            Try

                retorno = New ColecaoParcelas

                persistencia = New pParcela
                retornoPersistencia = persistencia.ConsultarPagamentos(dados)

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Pagamentos [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function ConsultarParcela(ByVal cid As Integer) As dParcelas
            Dim colecao As ColecaoParcelas
            Dim dados As dParcelas = New dParcelas()
            Dim persistencia As pParcela
            Dim retorno As dParcelas

            Try

                dados.cid = cid
                persistencia = New pParcela
                colecao = persistencia.Consultar(dados)

                If Not colecao Is Nothing Then
                    If colecao.Count > 0 Then
                        retorno = colecao(0)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em ConsultarParcela [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function ConsultarParcelasCliente(ByVal codCliente As Integer) As ColecaoParcelas

            Dim retorno As ColecaoParcelas
            Dim persistencia As pParcela
            Dim retornoPersistencia As ColecaoParcelas

            Try

                retorno = New ColecaoParcelas

                persistencia = New pParcela
                retornoPersistencia = persistencia.ConsultarParcelasCliente(codCliente)

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function ConsultarParcelasVencidas(ByVal codCliente As Integer) As ColecaoParcelas

            Dim retorno As ColecaoParcelas
            Dim persistencia As pParcela
            Dim retornoPersistencia As ColecaoParcelas

            Try

                retorno = New ColecaoParcelas

                persistencia = New pParcela
                retornoPersistencia = persistencia.ConsultarParcelasVencidas(codCliente)

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function Incluir(ByVal dados As dCrediario, ByVal dadosParcelas As ColecaoParcelas) As Integer

            Dim retorno As Integer
            Dim persistencia As pCrediario
            Dim persistenciaParcelas As pParcela

            Try

                Using ts As New TransactionScope

                    persistencia = New pCrediario
                    retorno = persistencia.Incluir(dados)

                    persistenciaParcelas = New pParcela()
                    For Each parcela As dParcelas In dadosParcelas
                        parcela.crediarioId = retorno
                        persistenciaParcelas.Incluir(parcela)
                    Next

                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Incluir Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function IncluirCrediario(ByVal dados As dCrediario) As Integer

            Dim retorno As Integer
            Dim persistencia As pCrediario

            Try

                Using ts As New TransactionScope

                    persistencia = New pCrediario
                    retorno = persistencia.Incluir(dados)

                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Incluir Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function IncluirParcela(ByVal dadosParcela As dParcelas) As Integer

            Dim retorno As Integer
            Dim persistenciaParcelas As pParcela

            Try

                Using ts As New TransactionScope

                    persistenciaParcelas = New pParcela()
                    persistenciaParcelas.Incluir(dadosParcela)

                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Incluir Parcela [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function Alterar(ByVal dados As dCrediario, ByVal dadosParcelas As ColecaoParcelas) As Integer

            Dim retorno As Integer
            Dim persistencia As pCrediario
            Dim persistenciaParcelas As pParcela

            Try

                Using ts As New TransactionScope
                    persistencia = New pCrediario
                    retorno = persistencia.Alterar(dados)

                    persistenciaParcelas = New pParcela()
                    For Each parcela As dParcelas In dadosParcelas
                        persistenciaParcelas.Alterar(parcela)
                    Next


                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Alterar Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function GravarPagamentoParcelas(ByVal dadosParcelas As ColecaoParcelas) As Integer

            Dim retorno As Integer
            Dim persistenciaParcelas As pParcela

            Try

                Using ts As New TransactionScope

                    persistenciaParcelas = New pParcela()
                    For Each parcela As dParcelas In dadosParcelas
                        persistenciaParcelas.IncluirPagamento(parcela)
                    Next

                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Gravar Pagamento Parcela [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function Renegociar(ByVal dados As dCrediario, ByVal dadosParcelas As ColecaoParcelas) As Integer

            Dim retorno As Integer
            Dim persistencia As pCrediario
            Dim persistenciaParcelas As pParcela

            Try

                Using ts As New TransactionScope
                    persistencia = New pCrediario
                    If dados.cid <> 0 Then
                        retorno = persistencia.Alterar(dados)
                    Else
                        retorno = persistencia.Incluir(dados)
                    End If

                    persistencia.ExcluirParcelas(dados)

                    persistenciaParcelas = New pParcela()
                    For Each parcela As dParcelas In dadosParcelas
                        If parcela.crediarioId = 0 Then
                            parcela.crediarioId = retorno
                        End If
                        persistenciaParcelas.Incluir(parcela)
                    Next

                    GravarLog(dados.usuarioId, "Renegociação de dívida - valor " & dados.ValorTotal & " cliente: " & dados.clienteId)

                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Alterar Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function AlterarCrediario(ByVal dados As dCrediario) As Integer

            Dim retorno As Integer
            Dim persistencia As pCrediario

            Try

                Using ts As New TransactionScope
                    persistencia = New pCrediario
                    retorno = persistencia.Alterar(dados)

                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Alterar Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function AlterarControle(ByVal dados As dCrediario) As Integer

            Dim retorno As Integer
            Dim persistencia As pCrediario

            Try

                Using ts As New TransactionScope
                    persistencia = New pCrediario
                    retorno = persistencia.AlterarControle(dados)

                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Alterar Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function Excluir(ByVal dados As dCrediario) As Integer

            Dim retorno As Integer
            Dim persistencia As pCrediario

            Try

                Using ts As New TransactionScope
                    persistencia = New pCrediario
                    retorno = persistencia.Excluir(dados)

                    GravarLog(dados.usuarioId, "Exclusão de crediário - valor " & dados.ValorTotal & " cliente: " & dados.clienteId)

                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Excluir Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function CorrigirParcelas() As Integer

            Dim retorno As Integer
            Dim persistencia As pParcela

            Try

                Using ts As New TransactionScope
                    persistencia = New pParcela
                    retorno = persistencia.Corrigir()
                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro ao corrigir parcelas [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

    End Class

End Namespace
